File: Computer_Science/Common_Algorithms/Sorting/Bubble_Sort/Bubble_Sort.py
def bubble_sort(arr):
    n = len(arr)
    swapped = True
    while swapped:
        swapped = False
        for i in range(1, n):
            if arr[i - 1] > arr[i]:
                # TIL
                arr[i - 1], arr[i] = arr[i], arr[i - 1]
                # temp = arr[i - 1]
                # arr[i - 1] = arr[i]
                # arr[i] = temp
                swapped = True
        n -= 1
    return arr


def main():
    arr = [37, 12, 89, 5, 61, 23, 8, 45, 72, 16]
    sorted_arr = bubble_sort(arr)
    for num in sorted_arr:
        print(num)


main()

File: quantinium/Computer_Science/Common_Algorithms/Sorting/Bubble_Sort/Bubble_Sort.c
#include <stdbool.h>
#include <stdio.h>

void swap(int *x, int *y) {
  int temp = *x;
  *x = *y;
  *y = temp;
}

// OPTIMIZED BUBBLE SORT
void bubble_sort(int *arr, size_t size) {
  bool swapped = true;
  size_t n = size;
  while (swapped) {
    swapped = false;
    for (size_t i = 1; i < n; i++) {
      if (arr[i - 1] > arr[i]) {
        swap(&arr[i - 1], &arr[i]);
        swapped = true;
      }
    }
    n--; // Optimization
  }
}

int main() {
  int array[10] = {37, 12, 89, 5, 61, 23, 8, 45, 72, 16};
  size_t size = sizeof(array) / sizeof(array[0]);

  bubble_sort(array, size);

  printf("After Sorting the array: \n");
  for (size_t i = 0; i < size; i++) {
    printf("%d ", array[i]);
  }
}

// SECOND IMPLEMENTATION
// void bubble_sort(int *arr, size_t size) {
//   bool swapped = true;
//   while (swapped) {
//     swapped = false;
//     for (int i = 1; i < size; i++) {
//       if (arr[i - 1] > arr[i]) {
//         swap(&arr[i - 1], &arr[i]);
//         swapped = true;
//       }
//     }
//   }
// }

// FIRST IMPLEMENTATION
// void swap(int *x, int *y) {
//   int temp = *x;
//   *x = *y;
//   *y = temp;
// }
//
// void bubble_sort(int *array, size_t size) {
//   for (int i = 0; i < size; i++) {
//     for (int j = 0; j < size - i - 1; j++) {
//       if (array[j] > array[j + 1]) {
//         swap(&array[j], &array[j + 1]);
//       }
//     }
//   }
// }
